src/nnvm/graph_dump.cc

```diff
@@ -0,0 +1,83 @@
+    /*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ * \file graph_dump.cc
+ * \brief Utilities to introspect and print the execution graph
+ * \author Pedro Larroy
+ */
+
+#include "./graph_dump.h"
+#include <vector>
+#include <string>
+#include "nnvm/graph.h"
+
+using std::vector;
+using std::string;
+using namespace std;
+
+using namespace nnvm;
+
+namespace {
+
+std::string unnamed(const std::string &s) {
```

Comment:
`unnamed` name does not really fit to what the purpose of this function is. Also, the logic here makes your graph potentially incorrect if a person used the same name (or the same lack of name) for multiple nodes (which is unfortunately allowed by MXNet).
What I would prefer is a function named something like `DeduplicateName` where you would check whether the same name was not used before for different Node and add `#number` to the name (adding `unnamed` or better `_unnamed_` would be fine there too).

Reply:
Thanks, makes sense.

Reply:
Should I add a static atomic counter for that for example?

Reply:
From what I tested most of the nodes have names from operators or variables, and head gradients I added names which makes the graph much better.

Reply:
I don't think you need an atomic counter for this - just a map that maps the `std::string` to list of `Node*` would be enough to differentiate whether your name is unique or not and what number you need to add to it.
There is nothing (unfortunately) in MXNet preventing you from naming your nodes or variables the same in the same graph. I agree that head grads should have a name, I actually did similar change in my pointwise fusion PR.

Reply:
Doesn't it do the same effect with less memory usage to what I proposed just for the purpose of adding a number?

Reply:
I guess I'm not sure how a counter would help you solve the issue of separate nodes that have the same name.

Reply:
Ok I think I got now your intention. I will make the suggested changes.